ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/core/WsAgentWebSocketConnectionTracker.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.ide.core;
+
+import static java.util.Collections.emptyMap;
+import static org.eclipse.che.api.core.model.workspace.runtime.ServerStatus.RUNNING;
+import static org.eclipse.che.api.workspace.shared.Constants.SERVER_WS_AGENT_HTTP_REFERENCE;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_HEARTBEAT;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_PERIOD;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_SUBSCRIBE;
+import static org.eclipse.che.api.workspace.shared.Constants.WS_AGENT_TRACK_CONNECTION_UNSUBSCRIBE;
+import static org.eclipse.che.ide.api.jsonrpc.Constants.WS_AGENT_JSON_RPC_ENDPOINT_ID;
+import static org.eclipse.che.ide.core.StandardComponentInitializer.STOP_WORKSPACE;
+import static org.eclipse.che.ide.util.browser.BrowserUtils.reloadPage;
+
+import com.google.gwt.user.client.Timer;
+import com.google.web.bindery.event.shared.EventBus;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestHandlerConfigurator;
+import org.eclipse.che.api.core.jsonrpc.commons.RequestTransmitter;
+import org.eclipse.che.api.core.model.workspace.Workspace;
+import org.eclipse.che.api.core.model.workspace.WorkspaceStatus;
+import org.eclipse.che.ide.CoreLocalizationConstant;
+import org.eclipse.che.ide.api.action.ActionManager;
+import org.eclipse.che.ide.api.app.AppContext;
+import org.eclipse.che.ide.api.workspace.event.WorkspaceStoppedEvent;
+import org.eclipse.che.ide.api.workspace.event.WsAgentServerRunningEvent;
+import org.eclipse.che.ide.api.workspace.event.WsAgentServerStoppedEvent;
+import org.eclipse.che.ide.api.workspace.model.MachineImpl;
+import org.eclipse.che.ide.api.workspace.model.ServerImpl;
+import org.eclipse.che.ide.bootstrap.BasicIDEInitializedEvent;
+import org.eclipse.che.ide.ui.dialogs.DialogFactory;
+
+/**
+ * Tracks web socket connection to control ws agent state. Notifies client side by {@link
+ * WsAgentServerStoppedEvent} when the connection is lost. Prompts to restart a workspace when
+ * connection is lost, but workspace is running.
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class WsAgentWebSocketConnectionTracker {
```

Comment:
Consider adding tests

Reply:
I need to test current implementation and going to add tests